### DISCUSSION TITLE
Subscriber Codeunits

### DISCUSSION THREAD

**Discussion by waldo1001** [microsoft/alguidelines#92] [BC Best Practices]
Subscribers are the new standard - but what are the best practices in subscriber-codeunit design?

Thinking about:
- bundling subscribers that belong together
- no business logic (small subscriber codeunit)
- single instance if possible

more?

**Comment by waldo1001** [comment]
@waldo1001 takes it

**Reply by TheDenSter** [reply]
> @waldo1001 takes it

What does this mean?

**Reply by TheDoubleH** [reply]
@TheDenSter It just means @waldo1001 will take the first stab at documenting

**Reply by waldo1001** [reply]
But @TheDenSter , if you want to describe this - please by all means ;-)

**Reply by TheDenSter** [reply]
I just wasnt sure what it meant so I asked

**Reply by waldo1001** [reply]
I wasn't able to really "assign" people to a topic, so I decided to use a comment .. 🤷‍♀️.

**Reply by aptMattKoe** [reply]
I've read it and like it :) thanks! 
But I'm not sure what I think about this Part: https://alguidelines.dev/bcbestpractices/performance/subscribercodeunits/#only-subscribe-when-necessary

Sure, only subscribe when its necassary but why to bind the subs before an validate in case of write that Code under the validate?
Or is only the example misleading?
In every case like this it also could be an separate call of a function instead of a subscription? Could be more complex or?

**Reply by waldo1001** [reply]
Misleading example, actually.  In this case, we're going to assume we can't just write the code on the OnValidate .. .

As said, in many case, we'll probably not be able to decide to "only subscribe when necessary" .. it's just: WHEN you can, manualbinding can improve performance 🤷‍♂️

I'm all ears for better use cases to remove the confusion ;-)

**Reply by aptMattKoe** [reply]
my first thoughts are going to events on the posting codeunits.
In that case, I want to add code in the middle of the process and not before/after that.

**Comment by kaspermoerch** [comment]
The pattern should include a naming convention for subscriber procedures.

For Integration and Table/Page Trigger events I propose: `{Object Type}_{Object Name}_{Event Name}`
For Table/Page Field events I propose: `{Object Type}_{Object Name}_{Event Name}_{Field Name}`

If one adheres to this naming pattern (combined with bundling events) it will enable the compiler to detect duplicate event subscribers. This will help minimize the number of event subscribers in an extension.

**Reply by waldo1001** [reply]
Hm, that can turn into a big discussion, I'm afraid ;-)

Yes, something should be said about naming conventions.  But I don't particularly agree with the proposed one. 
Personally, I don't think we should try all we can to avoid multiple subscribers for one event.  I'm more on the opposite side: isolate event subscriptions by functionality, create a codeunit per feature, including all subscribers for one feature (readabiliy), describing where it hooks into.. .

On top of that - imho - subscribers should not depend on the order/state/.. .  If they do, then there should be another event: OnAfterTheEventWeDependOn .. .

**Reply by BusinessCentralGeek** [reply]
So is @kaspermoerch proposal incompatible with your approach because you might subscribe multiple times to the same event at the same feature level/codeunit?

**Reply by waldo1001** [reply]
No, I would never subscribe twice from the same feature.  One feature (we call it "module") has one subscriber codeunit, with all subscriber to "outside the module".  Having multiple times the same subscribe in the same codeunit doesn't make much sense imho.  I would have multiple "oneliners" in that one subscribers, detailing why I'm calling out.

That said - in that case, I do admit that Kasper's naming convention would make a lot more sense (as you can't simply call it a certain methodname) 🤔.

**Reply by BusinessCentralGeek** [reply]
So.. do we have a deal? Only thing I will do it on PascalCase. That is:

[EventSubscriber(ObjectType::Codeunit, Codeunit::"Sales-Post", 'OnAfterCheckSalesDoc', '', false, false)]
would have '**CodeunitSalesPostOnAfterCheckSalesDoc**' as the procedure name.

Let´s see what others think about it.

**Reply by TheDenSter** [reply]
I don't agree with that naming convention at all, but this is one of those instances where you kind of have to pick the lesser of bad options. The eventsubscriber itself already defines the object, there shouldn't be a need to repeat that in the procedure name. I've come to just repeat the event name as the procedure name, and I keep those unique by creating a codeunit for all the subscriptions for a particular codeunit. So I'll have a "SalesPostSubs" codeunit for the events that I subscribe to from the "Sales-Post" codeunit, and the subscribers in there just have the same names as the events in the publisher codeunit. 

I do have to say that I also have an app out there where I have a codeunit called SalesPostABC.codeunit.al and in there I subscribe to various events around the sales posting process, so there are events that subscribe from multiple events. In that case I do use "SalesPostOnAfterCheckMandatoryFields" and "SalesPostYesNoOnBeforeConfirmSalesPost" and "InvoicePostBufferOnAfterInvPostBufferPrepareSales", but I'm not liking it because the objects are already defined in the event definition itself.

Ultimately it really doesn't matter what you call the event subscribers though. You should be able to find all reference by searching the procedure name of the event publisher.

The hardest part for me is to decide how to organize the subscribers themselves. Whether by object or by feature, at some point you get overlap and then it's even harder to find them. You subscribe to event x for feature A, and then you need the same event for feature B. Where are you going to put that procedure? Two subscribers? Piggy back on the one in feature A? Separate the subs in its own folder?

**Reply by waldo1001** [reply]
Two subscribers - always.  

Subscriber codeunit per feature makes most sense simply because you get a nice description where you're hooking into for this feature for free🤷‍♂️.  That's also why I described it that way in the guideline.

It's still easy to find both ways.  Where did I subscribe to the OnAfterPostSalesDoc?
![image](https://user-images.githubusercontent.com/12088142/156041883-0de8e35f-92f6-4606-9778-5d4894c1adb9.png)

Or - where did I hook into this "CSH"-feature?
![image](https://user-images.githubusercontent.com/12088142/156041980-7134dc67-9df1-46d1-8ccb-a83c065ef032.png)

**Reply by TheDenSter** [reply]
There's definitely more than one legitimate way to skin this cat, and the search is so good that you should have no problem finding stuff either way. For that same reason I kind of don't care to store physical object file in folders per feature. It's so easy to find things that it really makes no difference where the actual file itself is. I just turn on the crs extension in VSCode and let it create folders per object type. I'd just as much dump them all in a single 'src' folder. I'm using Ctrl+P to find objects more and more anyway so I'm slowly relying less on where the object itself is stored.

It looks like your approach is to basically to treat a feature as an app. I'm not sure if I'm on board with having three different subscribers into the same event in the same app. If I have to validate a field in a record, I'd have to get that record in each of the subs, which I could have done in a single read if I add the logic to a single sub in the whole app.

I'm still leaning toward a single sub codeunit for the 'base object' so I would just put all subs to the "Sales-Post" codeunit in a single "SalesPostSubs" codeunit, or even just call it "SalesPost" and have a comment at the top of the object that says something like "subscribers to the Sales-Post codeunit". I suppose you can still execute the actual business logic in other objects but for me, it 'feels better' if all subscribers for a particular codeunit are in one object per app. Not sure if I could write up why I feel that way, but it has something to do with coupling the subscriber logic to where you hook into the base app

**Reply by kaspermoerch** [reply]
It appears I put my hand down the hornets nest. Sorry about that.

The general purpose of my proposed naming convention is to minimize the number of subscribers for the same event. As I recall it when the event framework was first introduced, performance would decrease when the number of subscribers rise (I am not sure however if this is still the case).

In line with this I am like @TheDenSter leaning towards having one subscriber codeunit per base object.

**Reply by waldo1001** [reply]
:-).  I guess it's just a matter of opinion. 
Performance drops indeed .. but not **that** much to ignore other disadvantages it comes with.. .  (or at least in my opinion).

What we need to know in a lot of different cases is: "where does a feature hook into?" or better yet, when problem solving: "do we hook into everything we need to hook into?".

I really wonder how you will answer that with a search-string, searching through a series of sub-codeunits-per-object, relying on method names?  Affixes?  I don't know - hard to answer for sure.  While now, for me, it's just there in one, small, readable, feature-focused file.

On top of that, depending on the feature, I can "play" with single-instance, I can do variable persistance, isolated for this feature (upcoming pattern), I can implement manual binding much better (because all I need to look at is the functionality of my feature), ... .  I can treat the event subscribers by focusing on the feature, not the entire app.

**Reply by kaspermoerch** [reply]
> What we need to know in a lot of different cases is: "where does a feature hook into?" or better yet, when problem solving: "do we hook into everything we need to hook into?".

I definitely see and agree with the advantages of being able to isolate all code related to a feature. Doing it like that also keeps cohesion high and coupling low. In a scenario like that the procedure name of the event subscriber really doesn't matter.

So I guess for variations of this pattern a naming convention makes sense, but maybe not enough to be able to actually describe it well enough to be included in the pattern.

**Reply by waldo1001** [reply]
I was thinking - would it make sense to create 2 guidelines - both describing the pros and cons? 🤔.

I mean, I didn't come into a situation that I cared about subscriber-order - and since we are implementing the [generic method pattern](https://alguidelines.dev/docs/patterns/generic-method-pattern/), there is always some event we can hook into to maintain the order if that would be necessary. 

But that doesn't mean that there are never situations/apps where we would care 🤔.  And your approach IS handling that.

**Reply by kaspermoerch** [reply]
I'm a bit torn. I think patterns and best practices should promote feature based development. 

My suggested approach has some monolithic vibes over it, but having guidelines on how to do things right (even if you are taking a more monolithic approach) might be a good nudge on way to feature based development.

**Reply by waldo1001** [reply]
"feature based development" - I like that title 🤔.
Come to think of it, in my company, we kind of like to think that way in many areas:
- how to organize files
- how to manage subscribers
- how to manage affixes
- how to manage apps (yep, the smaller, the better)
- how to manage APIs (set of APIs per feature)
- how to manage telemetry signals (document per feature, numbering per feature, add featurename in customdimensions, ...)
- ...

But never consciously called it "feature based development/thinking" though 🤔

**Reply by BusinessCentralGeek** [reply]
> :-). I guess it's just a matter of opinion. Performance drops indeed .. but not **that** much to ignore other disadvantages it comes with.. . (or at least in my opinion).
> 

[https://www.youtube.com/watch?v=AwWWUiM1g8Y&list=PLihbooUf2JIxRhQbzJpQdsbhiyz8bFgSs&index=6](url)
Stefano´s video for best practices also tell us to have the least possible number of subscribers in a codeunit. So the "feature based development" (I also like it) could be the best approach.

![image](https://user-images.githubusercontent.com/92911762/156404196-5ebe9474-eb5f-4b33-a648-024b342445d6.png)

**Comment by TheDenSter** [comment]
You really approach features as mini-apps, apps within apps, applets, app-ushkas haha. I see the sense in that looking at it from a pov of organizing the objects but I'm just not sold on it. I've seen a bunch of apps that do that and it can get really messy that way. Then you see the dozens and dozens of apps that companies throw into extension management, and you think about all the features/applets inside of those. Just makes me shiver with all the 'stuff' that is in there.

**Reply by git-no** [reply]
It will improve maintainability to have only one subscriber function per event publisher.
Also the event subscriber codeunit should be single instance (because of performance reason => initialization) and have a small foot print of code within the functions. Within the subscriber function call the processing functions, can also be in separate objects of modules, e.g.:

```al
codeunit 2037325 "Setup Subs"
{
  SingleInstance = true;

  var
    KeyWordReportLbl: Label 'LT,Distri,Technical,Functional,Reports', Locked = true;  // Labels as global var

  [EventSubscriber(ObjectType::Codeunit, Codeunit::"Manual Setup", 'OnRegisterManualSetup', '', false, false)]
  local procedure OnRegisterManualSetup(sender: Codeunit "Manual Setup")
  var
      Object1: Codeunit "Object 1";  // objects as local because the object global vars are not reset, because this codeunit is single instance
      Object2: Codeunit "Object 2";
  begin
      Object1.Method(sender);
      Object2.Method(sender);
  end;
}
```

The subscriber functioning name should be handy, I use the same name as the event subscriber, except for fields where the field name is added to the subscriber function name.

The worst would be to have multiple subscriber functions of one event publisher and to have these subscriber functions  in multiple files.

**Comment by extranrollo** [comment]
Question:

We have the chance to use a Subscriber to a field like

``` 
[EventSubscriber(ObjectType::Table, Database::"Sales Header", OnAfterValidateEvent, "Sell-to Customer No.", false, false)]
local procedure SalesHeader_OnAfterValidateEvent_SellToCustomerNo(var Rec: Record "Sales Header"; var xRec: Record "Sales Header"; CurrFieldNo: Integer)
var
  MySalesManagement: Codeunit "My Sales Management";
begin
  MySalesManagement.DoSomeThing(Rec);
end; 
```

or in a tableextension

```
modify("Sell-to Customer No.")
{
   trigger OnAfterValidate()
   var
     MySalesManagement: Codeunit "My Sales Management";
   begin
     MySalesManagement.DoSomeThing(Rec);
   end;
} 
```

How does this have effect to performance. As understood the EventSubscribers do have a effect, because they have to be loaded in memory on execution.

But when i have it directly in the extension, i imagine that the code is already loaded by reading the object and so it could be faster this way. 

How do you see that?
